src/tests/column_options/auto_increment.rs

```diff
@@ -0,0 +1,175 @@
+use crate::*;
+
+test_case!(auto_increment, async move {
+    use Value::*;
+
+    let test_cases = vec![
+        (
+            "CREATE TABLE Test (id INTEGER AUTO_INCREMENT NOT NULL, name TEXT)",
+            Ok(Payload::Create),
+        ),
+        (
+            r#"INSERT INTO Test (name) VALUES ('test1')"#,
+            Ok(Payload::Insert(1)),
+        ),
+        (
+            r#"SELECT * FROM Test"#,
+            Ok(select!(
+            id  | name
+            I64 | Str;
+            1    "test1".to_owned())),
+        ),
+        (
+            r#"INSERT INTO Test (name) VALUES ('test2'), ('test3')"#,
+            Ok(Payload::Insert(2)),
+        ),
+        (
+            r#"SELECT * FROM Test"#,
+            Ok(select!(
+            id  | name
+            I64 | Str;
+            1    "test1".to_owned();
+            2    "test2".to_owned();
+            3    "test3".to_owned())),
+        ),
+        (
+            r#"INSERT INTO Test (name, id) VALUES ('test4', NULL)"#,
+            Ok(Payload::Insert(1)),
+        ),
+        (
+            r#"SELECT * FROM Test"#,
+            Ok(select!(
+            id  | name
+            I64 | Str;
+            1    "test1".to_owned();
+            2    "test2".to_owned();
+            3    "test3".to_owned();
+            4    "test4".to_owned())),
+        ),
+        (
+            r#"INSERT INTO Test (name, id) VALUES ('test5', 6)"#,
+            Ok(Payload::Insert(1)),
+        ),
+        (
+            r#"SELECT * FROM Test"#,
+            Ok(select!(
+            id  | name
+            I64 | Str;
+            1    "test1".to_owned();
+            2    "test2".to_owned();
+            3    "test3".to_owned();
+            4    "test4".to_owned();
+            6    "test5".to_owned())),
+        ),
+        (
+            r#"INSERT INTO Test (name) VALUES ('test6')"#,
+            Ok(Payload::Insert(1)),
+        ),
+        (
+            r#"SELECT * FROM Test"#,
+            Ok(select!(
+            id  | name
+            I64 | Str;
+            1    "test1".to_owned();
+            2    "test2".to_owned();
+            3    "test3".to_owned();
+            4    "test4".to_owned();
+            6    "test5".to_owned();
+            5    "test6".to_owned())),
```

Comment:
This might be `7` instead of `5`?
`6` is manually inserted then auto increment value also should be updated.

Reply:
Why should it be? It isn't an *auto* value?

Reply:
Say for example you are only up to 5 and someone inserts 100, should it go to 6, 100 or stay on 5?

Reply:
someone insert 100, then the next value need to become 101.

Ok, I'm basically for not distinguishing values which are whether inserted by user or system(auto-increment).
This is also approaches of both MySQL and SQLite.

Current implementation handle values differently, for auto increment value, it ignores user entered value.

Reply:
The functionality for the different pr was this,
We decided to generate error when redundant value comes to `UNIQUE` column.
For example,

```sql
CREATE TABLE Test (id INTEGER UNIQUE AUTO_INCREMENT);

INSERT INTO Test VALUES (NULL);
INSERT INTO Test VALUES (NULL);
INSERT INTO Test VALUES (NULL);
-- 1, 2, 3

INSERT INTO Test VALUES (10);
-- 1, 2, 3, 10

INSERT INTO Test VALUES (NULL);
INSERT INTO Test VALUES (NULL);
-- 1, 2, 3, 10, 11, 12

INSERT INTO Test VALUES (8);
-- 1, 2, 3, 10, 11, 12, 8
INSERT INTO Test VALUES (NULL);
-- 1, 2, 3, 10, 11, 12, 8, 9

INSERT INTO Test VALUES (NULL);
-- Error! (for this pr)

-- using the future pr, we can provide support not to generate error
-- then, 1, 2, 3, 10, 11, 12, 8, 9, 13
```

> Note that this sort of feature would add notable weight to inserts.

For this, not special work is required.

MySQL
```sql
create table Test(id integer primary key auto_increment, title varchar(100));
insert into Test(id, title) values(1, "Hello");
insert into Test(id, title) values(NULL, "Good");
insert into Test(id, title) values(100, "Good");
insert into Test(id, title) values(NULL, "Good");

-- Your code here!

update Test set id = 200 where id = 100;

insert into Test(id, title) values(NULL, "Good");

delete from Test where id = 201;

insert into Test(id, title) values(NULL, "Good");
select * from Test;
```

Then, execution result is this below
| id | title |
| ----- | ----- |
| 1 | Hello |
| 2 | Good |
| 200 | Good |
| 101 | Good |
| 202 | Good |

I think MySQL's strategy is quite straightfoward.
It checks and store the biggest value when `INSERT` and `UPDATE` occurs which is for avoiding redundant number input occurs.
However, it does not care whatever `DELETE` does. because `DELETE` cannot make bigger value, it can only remove values.

How do you think?

Reply:
I, personally, don't like that.

I am familiar with MSSQL's method which is what I've implemented (we just don't have `RESEED` yet) and the way I've implemented it is more efficient than the ways you are mentioning.

These will require checking each inserted value into any AI column.

Under my method we do:
1. Filter AI columns - `n = columns` iters
2. Filter `NULL` rows for count - `n = AI_columns * rows` iters
3. Get from sled - `n = AI_columns` times
4. Add the filter count to the gets  - `n = AI_columns` times
5. Insert into sled - `n = AI_columns` times 
6. Check if rows are `NULL` - `n = AI_columns * rows` iters
7. If `NULL`, set value (and increment) - `n = rows(NULL)` iters

Using that method we also do:
- Sort rows - `n = AI_columns * rows * ?` iters
- Insert into sled - `n = AI_columns` times
or:
- Filter to rows that aren't `NULL` - `n = AI_columns * rows` iters
- Insert into sled - `n = AI_columns * rows(NOT NULL)` times

And if we want to keep highest value:
- Sort rows - `n = AI_columns * rows * ?` iters
- Get from sled - `n = AI_columns` times
- Compare highest and sled - `n = AI_columns` times
- Insert into sled - `n = AI_columns * ?` times
or:
- Filter to rows that aren't `NULL` - `n = AI_columns * rows` iters
- Get from sled - `n = AI_columns * rows(NOT NULL)` times
- Compare with sled - `n = AI_columns * rows(NOT NULL)` times
- Insert into sled - `n = AI_columns * rows(more than sled)` times

Results per insert would be the same, only difference would be consecutive inserts.

If one wanted to enable changing per insert, notably more processing would be required.

Reply:
My opinion is that, for an AI column, you mostly just leave the database to do it unless you want to manually intervene, in which case you can also run `RESEED`.

If for some reason you were regularly inserting values into an AI column, yes, I can see why that form is preferrable, this is why I think it would be better to have this as an optional mechanic.

Reply:
Oh! ok btw I think now I can understand... I was wondering what causes some kind of communication miss in here.
I only considered the implementations of MySQL and SQLite, they are two main examples of `AUTO_INCREMENT` and `AUTOINCREMENT`.

I didn't care about behavior of MSSQL at all, and that is what you are familiar with.

Ok, so you are considering to implement subset of MSSQL `IDENTITY` behavior using `AUTO_INCREMENT` keyword?

Reply:
Yup, a bit of a mix of the behaviours of various database but leaning towards MSSQL's `IDENTITY` for the fast method.
I've left it really open to be able to add optional implementations that expand to be like MySQL or SQLite though :-)

Reply:
Good, then can you share some brief design specification of this auto incrementation? good place would be in discussion tab.

I think that current most of our miss communication was because we simply had different idea about the AI feature itself.
If it is implementing simply `MySQL`'s one or `SQLite`'s one, then it's just ok to see the document from them.
Or if you use `IDENTITY` for this auto increment feature, then it would be also ok to simply read the document from MSSQL.

> a bit of a mix of the behaviours of various database 

However, it looks like that you are doing something different. Then I cannot figure out what's the feature which is supposed be implemented, it means that I cannot review the codes because I don't know the spec.

Reply:
https://github.com/gluesql/gluesql/discussions/194